File: day05/day5.py
import re
from collections import defaultdict
from math import copysign as sign

def buildVents(data, blockDiag=True):
    vents = defaultdict(lambda:0)

    for (a, b, x, y) in data:
        if blockDiag and a != x and b != y: continue
        i, j = a, b
        xDir = int(sign(1, x-a)) if x != a else 0
        yDir = int(sign(1, y-b)) if y != b else 0
        while i != x or j != y:
            vents[(i,j)] += 1
            i, j = i+xDir, j+yDir
        vents[(i,j)] += 1
    return len([v for v,a in vents.items() if a > 1])

def part1(data):
    return buildVents(data)

def part2(data):
    return buildVents(data, False)
    
def main():
    with open('input.txt') as f:
        data = [[int(x) for x in ent] for ent in re.findall(r'(\d+),(\d+) -> (\d+),(\d+)', f.read().strip())]

    print(f'Part 1: {part1(data)}')
    print(f'Part 2: {part2(data)}')

if __name__ == "__main__":
    main()

File: day05/day5.js
const { readFileSync } = require("fs");

const buildVents = (data, blockDiag = true) => {
  vents = {};
  data.forEach(([a, b, x, y]) => {
    if (blockDiag && a != x && b != y) return;
    let i = a,
      j = b;
    const xDir = Math.sign(x - a);
    const yDir = Math.sign(y - b);
    while (i != x || j != y) {
      const k = `${i},${j}`;
      if (!(k in vents)) vents[k] = 0;
      vents[k]++;
      i += xDir;
      j += yDir;
    }
    const k = `${i},${j}`;
    if (!(k in vents)) vents[k] = 0;
    vents[k]++;
  });
  let hot = 0;
  for (const [_, laps] of Object.entries(vents)) {
    if (laps > 1) hot++;
  }
  return hot;
};

const part1 = (data) => {
  return buildVents(data);
};
const part2 = (data) => {
  return buildVents(data, (blockDiag = false));
};

const input = readFileSync("./input.txt", "utf-8");
const data = [...input.matchAll(/(\d+),(\d+) -> (\d+),(\d+)/g, input)].map(
  (match) => match.splice(1, 5).map((i) => parseInt(i))
);

console.log(`Part 1: ${part1(data)}`);
console.log(`Part 2: ${part2(data)}`);
